File: frontend/src/mocks/pong/simulation/step.ts
import {
  resolvePaddleCollisions,
  resolveSideBorders,
} from "./collisions";
import { moveBall } from "./motion";
import { decayPaddleVelocities } from "./paddles";
import { resolveGoals } from "./scoring";
import { isMatchFinished } from "./state";
import type { SimState } from "./types";

/**
 * Executa um passo da simulação na ordem:
 * decaimento das paddles, movimento da bola, bordas, paddles e gols.
 *
 * @param state - Estado da simulação
 * @param deltaSeconds - Tempo decorrido desde o último passo, em segundos
 */
export function stepSimulation(state: SimState, deltaSeconds: number): void {
  if (isMatchFinished(state)) return;

  decayPaddleVelocities(state, deltaSeconds);
  moveBall(state, deltaSeconds);
  resolveSideBorders(state);
  resolvePaddleCollisions(state);
  resolveGoals(state);
}

File: backend/app/games/pong/step.py

from .collisions import resolve_paddle_collisions, resolve_side_borders
from .motion import move_ball
from .paddles import decay_paddle_velocities
from .scoring import resolve_goals
from .state import SimState, is_match_finished

def step_simulation(state: SimState, delta_seconds: float) -> None:
    if is_match_finished(state):
        return

    decay_paddle_velocities(state, delta_seconds)
    move_ball(state, delta_seconds)
    resolve_side_borders(state)
    resolve_paddle_collisions(state)
    resolve_goals(state)
